js/web/lib/wasm/jsep/webgpu/ops/3rd-party/matmul_packed_webgpu.ts

```diff
@@ -452,39 +455,89 @@ export const createMatmulProgramInfo =
 
       const dataType = tensorTypeToWsglStorageType(inputs[0].dataType);
       const components = isVec4 ? 4 : 1;
-      const A = inputVariable('a', inputs[0].dataType, [...outerDimsA, dimAOuter, dimInner / components], components);
-      const B = inputVariable('b', inputs[1].dataType, [...outerDimsB, dimInner, dimBOuter / components], components);
-      const output =
-          outputVariable('result', inputs[0].dataType, [batchSize, dimAOuter, dimBOuter / components], components);
+
+      const aShapeTemp = [...outerDimsA, dimAOuter, dimInner / components];
+      const enableAShapesUniforms = enableShapesUniforms(aShapeTemp.length);
+      const aShapeOrRank = enableAShapesUniforms ? aShapeTemp.length : aShapeTemp;
+
+      const bShapeTemp = [...outerDimsB, dimInner, dimBOuter / components];
+      const enableBShapesUniforms = enableShapesUniforms(bShapeTemp.length);
+      const bShapeOrRank = enableBShapesUniforms ? bShapeTemp.length : bShapeTemp;
+
+      const outputShapeTemp = [batchSize, dimAOuter, dimBOuter / components];
+      const enableOutputShapesUniforms = enableShapesUniforms(outputShapeTemp.length);
+      const outputShapeOrRank = enableOutputShapesUniforms ? outputShapeTemp.length : outputShapeTemp;
+
+      const A = inputVariable('a', inputs[0].dataType, aShapeOrRank, components);
+      const B = inputVariable('b', inputs[1].dataType, bShapeOrRank, components);
+      const output = outputVariable('result', inputs[0].dataType, outputShapeOrRank, components);
       variables.push(A);
       variables.push(B);
       variables.push(output);
-      const inputVariables = [A, B];
+      const inputVariables = [batchDims, A, B];
```

Comment:
This is used by batchDims.offsetToIndices('u32(batch)'). Infact this is not input, it uses the logic of IndiceHelper to implement offsetToIndices, an example is like this:
```
  fn o2i_batchDims(offset: u32) -> vec3<u32> {
    var indices: vec3<u32>;
    var current = offset;
    
    let dim0 = current / uniforms.batchDims_strides[0];
    let rest0 = current % uniforms.batchDims_strides[0];
    indices[0] = dim0;
    current = rest0;
    
    let dim1 = current / uniforms.batchDims_strides[1];
    let rest1 = current % uniforms.batchDims_strides[1];
    indices[1] = dim1;
    current = rest1;
    indices[2] = current;
    return indices;
  }
```

Reply:
'batchDims' is just used as an IndiceHelper . It's not supposed to be declared. I think the problem is we missing a `registerVariableUniform` function. Do you think it's better to change common.ts like below: In this case, we don't need the `uniformOnly` flag. And existed code can keep consistent.
```
--- a/js/web/lib/wasm/jsep/webgpu/ops/common.ts
+++ b/js/web/lib/wasm/jsep/webgpu/ops/common.ts
@@ -731,8 +731,7 @@ class ShaderHelperImpl implements ShaderHelper {
   `;
   }

-  private declareVariable(variable: IndicesHelper, bindingIndex: number): string {
-    this.indicesHelpers.push(variable);
+  registerVariableUniform(variable: IndicesHelper) {
     if (variable.rank !== 0) {
       if (variable.shape.startsWith('uniforms.')) {
         this.uniforms.push({name: variable.shape.replace('uniforms.', ''), type: variable.type.indices});
@@ -741,6 +740,11 @@ class ShaderHelperImpl implements ShaderHelper {
         this.uniforms.push({name: variable.strides.replace('uniforms.', ''), type: variable.type.indices});
       }
     }
+  }
+
+  private declareVariable(variable: IndicesHelper, bindingIndex: number): string {
+    this.indicesHelpers.push(variable);
+    this.registerVariableUniform(variable);
     const access = variable.usage === 'input' ? 'read' : 'read_write';
     const storageType = variable.type.storage;
     return `@group(0) @binding(${bindingIndex}) var<storage, ${access}> ${variable.name}: array<${storageType}>;`;

Reply:
Not sure if this is better. Because we first define an InputVariable, then we didn't include it when declareVariables

Reply:
`batchDims` is used as an internal IndicesHelper not an input/output variable. Maybe we shouldn't re-use the `inputVariable`  type and define another type like `variable` to avoid the name controversy. Or add an indicator in `inputVariable` like `internal`/`private` to say this variable is not to be used as an input. Personally, I don't like the `uniformOnly` flag since `shapeOrRank` is already used to indicate whether to use uniform or not. It's hard to understand the `uniformOnly`'s meaning here. Since it's my personal preference, I will leave it as open. Other parts looks good to me. Thanks.